### ISSUE TITLE
GWT compiler complains about missing imports

### ISSUE THREAD

**Issue by sebasbaumh** [gwtproject/gwt#9560]
**GWT version:** 2.8.2
**Browser (with version):** N/A
**Operating System:** Windows 7 64 bit
---
##### Description
With the latest GWT version 2.8.2 the GWT compiler complains about missing imports, which are referenced in a Java file, but actually not used in GWT.

I am reusing a library in GWT and have marked all functions using types not supported by GWT with the attribute `@GwtIncompatible`.
For older GWT versions like 2.8.1 this works fine even when including the imports of this unsupported types on top of the file.

For the latest GWT version 2.8.2 it shows that error:
```
[INFO] Compiling module de.test.Test
[INFO]    Tracing compile failure path for type 'de.test.Util'
[INFO]       [ERROR] Errors in 'de/test/Util.java'
[INFO]          [ERROR] Line 14: The import java.time cannot be resolved
```
As it worked before, I consider this a regression.

##### Steps to reproduce
Create a file `de/test/Util.java` using this import:
```
import java.time.*;
```

And a class with this member function:
```
	@GwtIncompatible("DateTimeFormatter")
	public static String formatISO8601DateTime(ZonedDateTime d)
	{
		DateTimeFormatter dtf = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss.SSSXX");
		return dtf.format(d);
	}
```

While compiling the class should be generally usable, though the function formatISO8601DateTime is not (which is intended by marking it incompatible with GWT).
But the compiler shows the error `The import java.time cannot be resolved`.

**Comment by sebasbaumh** [comment]
I further tested this issue and it seems like the error only occurs for wildcard imports like `import java.time.*;` but not for imports of individual classes like `import java.time.ZonedDateTime;`.

**Comment by niloc132** [comment]
Confirmed that this breaks in 2.8.2, but does not in 2.8.1.

**Comment by niloc132** [comment]
It looks like https://gwt-review.googlesource.com/c/gwt/+/19020 was responsible for this - adding support for Java 9 broke our ability to ignore these imports? Reverting that patch fixes this case again. @rluble any chance you can take a glance, at least point me in the right direction? I'll take a crack at getting this fixed later this week.

**Comment by kirillp** [comment]
This bug is floating for some reason, our build agents fail to build randomly with this error.
The class causing the error is annotated as GwtIgnore.

**Comment by rluble** [comment]
Giving some context here.

 In order to compile a java file GWT uses JDT to first parse the Java file and removes the GwtIncompatible members from the AST before handing it back to JDT to resolve types and continue compilation.

The removal of GwtIncompatible members is done by GwtIncompatiblePreprocessor.java followed by UnusedImportsRemover.  The latter remove named imports and is crucial because leaving a named import to a class that does not exist causes JDT to fail the compile. But that pass can not be done precisely as there is some ambiguity which is resolved in the compilation process. UnusedImportsRemover does not attempt to remove .* imports (nor static imports).

This failure mode can occur even reverting the patch if the package to be resolved is not part of the jre.

A simpler solution that might work is to *assume* that a sequence of lower case components in which the first is "java" is always a package and make isPackage to return true for that case instead of trying to resolve it.

A workaround for the problem is to use a named import (avoiding * imports is normally a recommended practice).

**Comment by sebasbaumh** [comment]
@rluble you are right, that a named import does avoid the issue, at least where it is feasible to "unroll" the wildcard imports.

Nevertheless it seems to be a regression and it would be good to fix it if possible.
I also tried to work around this issue by creating a super source containing the given package and a dummy class - without success. (I first thought it could be related to the package not being used in the GWT JRE emulation).

**Comment by rluble** [comment]
The simplest solution is the hack I proposed, just modify `com.google.gwt.dev.javac.JdtCompiler.isPackage`.

After the first two checks add the following:
```java
  // Assume that a slash separated name that starts with "java" and is all lowercase is a package.
  if (slashedPackageName.matches("java/[a-z/]+")) {
    packages.add(slashedPackageName);
    return true;
  }
```

could you test it and add a relevant test. I'd be happy to review it.

**Comment by gkdn** [comment]
This was working by accident before. Import with "star" simply doesn't work for GwtIncompatible; adding a workaround just for java package doesn't seem right unless this is affecting too many people (I doubt) and because it is a point release. Otherwise we should just leave it alone.

**Comment by sebasbaumh** [comment]
@gkdn: so if this was working by accident, maybe you could just document this behavior as it is like "do not use wildcard imports, only named imports".

@rluble I think this won't fix it as most of the packages we use are in namespaces other than `java` like `de/company/something`. So it may just make it more confusing ;-)

**Comment by rluble** [comment]
The previous behavior (removed in https://gwt-review.googlesource.com/c/gwt/+/19020) only cared about packages in the Jre (mostly starting with java.). The error you are pointing out is about importing "java.time.*". This one should go away. The errors importing de.company.something.* would already be there in 2.8.1.  The hack I proposed only addresses the regression. 

But I agree with @gkdn, the hack I proposed is only a last resort, if there are many breakages for existing applications that are costly to fix. It is definitely discouraged to use .* imports.

Closing as won't fix.
